### PR TITLE
feat(migrate-engine): start moving out of introspection-engine

### PR DESCRIPTION
**Pull request by jkomyno**
Closes https://github.com/prisma/schema-team/issues/375

**TODO**s after `prisma@4.9.0`:
- [x] https://github.com/prisma/prisma-private/issues/203

### CODE REVIEW

File: packages/internals/src/sendPanic.ts
Comment (jkomyno) [inline packages/internals/src/sendPanic.ts:49]: The `MigrateEngine` may throw Rust panics with the `LIFT_CLI` error area, rather than the previous `INTROSPECTION_CLI` error area thrown by the `IntrospectionEngine`

File: packages/internals/src/sendPanic.ts
Comment (jkomyno) [inline packages/internals/src/sendPanic.ts:35]: Turning a list of arguments to an object with named properties for ease of use. It'd be easy passing the arguments in the wrong sequence instead.

**Comment by jkomyno** [comment]
Integration test errors should be fixed by https://github.com/prisma/prisma-engines/pull/3495.

File: packages/cli/src/Doctor.ts
Comment (janpio) [inline packages/cli/src/Doctor.ts:90]: For later: Would `cwd` maybe be a better name for `projectDir` as well?

File: packages/internals/src/sendPanic.ts
Comment (janpio) [inline packages/internals/src/sendPanic.ts:25]: I think the old name of `engineVersion` (singular, not plural) was chosen because it is the version of _one_ engine. In theory all engines in the CLI can use different versions, and this highlights that this is specifically the version from that one engine file.

File: packages/internals/src/sendPanic.ts
Comment (janpio) [inline packages/internals/src/sendPanic.ts:34]: Not used before anyway so dead?

File: packages/internals/src/sendPanic.ts
Comment (janpio) [reply packages/internals/src/sendPanic.ts:49]: No chance of this now triggered for ME errors as well? (probably undertested area)

File: packages/internals/src/sendPanic.ts
Comment (janpio) [reply packages/internals/src/sendPanic.ts:34]: Ah, I see below.

File: packages/internals/src/sendPanic.ts
Comment (janpio) [reply packages/internals/src/sendPanic.ts:49]: Ah, I see this is only to get the database version in a safe way - not more.

File: packages/internals/src/sendPanic.ts
Comment (janpio) [inline packages/internals/src/sendPanic.ts:42]: Are we sure that was not used for something and e.g. submitted with the error report?

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [inline packages/migrate/src/MigrateEngine.ts:153]: Why does this now also take/need the schema?

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [inline packages/migrate/src/MigrateEngine.ts:159]: PR https://github.com/prisma/prisma-engines/pull/3487 got dismissed

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [inline packages/migrate/src/MigrateEngine.ts:327]: Now way that the code after this still makes it fail in some way that `isRunning=true` would be inaccurate?

File: packages/migrate/src/utils/getDatabaseVersionSafe.ts
Comment (janpio) [inline packages/migrate/src/utils/getDatabaseVersionSafe.ts:11]: When reading the code, this name threw me off, but I guess it makes sense to describe what it is doing.

File: packages/migrate/src/utils/getDatabaseVersionSafe.ts
Comment (janpio) [inline packages/migrate/src/utils/getDatabaseVersionSafe.ts:18]: Does this always start up a new Engine to make this query, or would it reuse an existing one if it already exists?

File: packages/migrate/src/MigrateEngine.ts
Comment (jkomyno) [reply packages/migrate/src/MigrateEngine.ts:159]: I had done that and reverted it in https://github.com/prisma/prisma/pull/16766#:~:text=%E2%80%A6-,392a6d2,-jkomyno%20added%20this, as apparently that PR alone doesn't fix this type mismatch problem in JSON-RPC between JS and Rust

File: packages/migrate/src/MigrateEngine.ts
Comment (jkomyno) [reply packages/migrate/src/MigrateEngine.ts:327]: This `isRunning` properties is implemented mirroring the same property in the IntrospectionEngine, and is updated in the same way. The idea is to first get the MigrateEngine on par with the features exposed by the IntrospectionEngine we need to make use of, and later verify possible improvements that could be done

File: packages/migrate/src/MigrateEngine.ts
Comment (jkomyno) [reply packages/migrate/src/MigrateEngine.ts:153]: The `getDatabaseVersion` RPC method had always required a schema. We didn't notice that the `getDatabaseVersion` recently added to MigrateEngine was missing an input argument because that method wasn't used in tests

File: packages/internals/src/sendPanic.ts
Comment (jkomyno) [reply packages/internals/src/sendPanic.ts:42]: I had asked both Joel and Tom before removing `getDatabaseDescription` (which isn't implemented for MigrateEngine and isn't used anywhere else). The rationale is that we haven't made use of it in a long time (Tom never used it, Matthias did once or twice), and if we can remove legacy non-user-facing stuff, that's "always the best option"

File: packages/cli/src/Doctor.ts
Comment (jkomyno) [reply packages/cli/src/Doctor.ts:90]: That's a potential choice, yeah

File: packages/internals/src/sendPanic.ts
Comment (janpio) [reply packages/internals/src/sendPanic.ts:42]: This referred to the resulting `sqlDump` - if we should not be using that to report to the error report. Might very well be that this system is actually set up to accept that and might benefit from doing so.

File: packages/internals/src/sendPanic.ts
Comment (jkomyno) [reply packages/internals/src/sendPanic.ts:25]: I'm glad we discussed this then! I'm creating a private issue about it

**Comment by Jolg42** [comment]
Note from Alberto: we need Tom to unblock the progress on this PR on engines side.

File: packages/migrate/src/commands/DbPull.ts
Comment (Jolg42) [inline packages/migrate/src/commands/DbPull.ts:257]: I think that this shows a breaking change in our public API here before we had the error code `P4001` and now it disappears (changed to P1003)
We need to find another way and keep the old codes or go for a major version bump.

Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts:257]: Answered here: https://prisma-company.slack.com/archives/C04HZ2EQS65/p1673446997564369?thread_ts=1673440456.031899&cid=C04HZ2EQS65

Comment (Jolg42) [reply packages/migrate/src/commands/DbPull.ts:257]: TODO:
We now we have 2 cases
- Database is empty
- Database does not exist
We need to check if we have tests for both in introspection

Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts]: I've added human-friendly error support for both `P4001` and `P1003`. Please check https://github.com/prisma/prisma/pull/16766/commits/9c820ae3348de82991defb068750189426bbfaa5

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (Jolg42) [inline packages/migrate/src/__tests__/DbPull.test.ts:817]: Why is this skipped now?

File: pnpm-lock.yaml
Comment (Jolg42) [inline pnpm-lock.yaml:232]: Diff in the pnpm lock is unexpected, did you install this engine before it landed in main?

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:421]: As of yesterday evening, not anymore. We can get rid of it

File: pnpm-lock.yaml
Comment (jkomyno) [reply pnpm-lock.yaml:232]: Yes, I had tried it out before it landed in main to see whether all tests passed. I can revert https://github.com/prisma/prisma/pull/16766/commits/fcbd3bcf9cf70db9125b12ea70b59739d2fc0243 and merge `main` back to this branch

File: packages/migrate/src/utils/test-handlePanic.ts
Comment (Jolg42) [inline packages/migrate/src/utils/test-handlePanic.ts:15]: Could you use the one from `getDatabaseVersionSafe.ts` ?

Comment (jkomyno) [reply packages/migrate/src/utils/test-handlePanic.ts:15]: I see that the same test in `packages/cli` uses the literal `prismaEngineVersionHash` (https://github.com/prisma/prisma/blob/dba8c89d305d91b7cdd6cac6fea49f5192d39af3/packages/cli/src/utils/test-handlePanic.ts#L8), I may just use the same in `packages/migrate`

File: packages/migrate/src/types.ts
Comment (Jolg42) [inline packages/migrate/src/types.ts:309]: It's missing
```
interface IntrospectionWarningsTopLevelItemNameIsADupe extends IntrospectionWarning {
  code: 20
  affected: AffectedTopLevel[]
}
```

Can you add the recent changes from this commit?
https://github.com/prisma/prisma/commit/c2d32eea77e963778ba0b98cc3557ad0de62a5b7#diff-a3ca29a928d7e7c324513f5239c93d68b07cbd190c678587efc3eb2a93c32d82

Comment (Jolg42) [reply packages/migrate/src/types.ts:309]: Note: It's the only change since this PR was opened

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:817]: Heck, this was unintentional, I had CMD+F'd the "skip" string in https://github.com/prisma/prisma/pull/16766/files without realizing `DbPull.test.ts` was compressed due to the diff being too large

File: packages/migrate/src/utils/test-handlePanic.ts
Comment (Jolg42) [reply packages/migrate/src/utils/test-handlePanic.ts:15]: Actually this file is not even used anywhere, it's a small script that someone can run locally (not part of tests)

So anything works I guess, but let's still remove `const getDatabaseVersionSafe = () => Promise.resolve(undefined)` so we don't forget later

File: pnpm-lock.yaml
Comment (Jolg42) [reply pnpm-lock.yaml]: It's probably fine in the end but you will get a conflict in this PR once a new engine version lands

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:817]: Thankfully the test passes anyway

File: packages/migrate/src/utils/test-handlePanic.ts
Comment (jkomyno) [reply packages/migrate/src/utils/test-handlePanic.ts:15]: Why would we remove `const getDatabaseVersionSafe = () => Promise.resolve(undefined)`? It's purposedly a mocked function

File: pnpm-lock.yaml
Comment (jkomyno) [reply pnpm-lock.yaml]: I have already done the steps described in my previous message with no conflicts whatsoever :)

File: packages/migrate/src/types.ts
Comment (jkomyno) [reply packages/migrate/src/types.ts:313]: Done

File: packages/cli/schema.prisma
Comment (Jolg42) [inline packages/cli/schema.prisma:1]: This schema file at the root of the CLI package is unexpected to me.
Could you remove it? (Is it needed for anything?)

File: packages/migrate/src/__tests__/introspection/introspection.test.ts
Comment (Jolg42) [inline packages/migrate/src/__tests__/introspection/introspection.test.ts:54]: Still needed?

File: packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts
Comment (Jolg42) [inline packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:314]: Still needed?

File: packages/internals/src/__tests__/introspection/introspection.test.ts
Comment (Jolg42) [inline packages/internals/src/__tests__/introspection/introspection.test.ts:67]: Still needed?

File: packages/cli/src/__tests__/artificial-panic.test.ts
Comment (janpio) [inline packages/cli/src/__tests__/artificial-panic.test.ts:59]: If now introspection errors also report `LIFT_CLI`, do we still have enough information in the error reporting tool to quickly place where the error comes from?

File: packages/internals/src/__tests__/introspection/introspection.test.ts
Comment (Jolg42) [reply packages/internals/src/__tests__/introspection/introspection.test.ts:67]: I think it is actually needed here

File: packages/cli/src/__tests__/artificial-panic.test.ts
Comment (jkomyno) [reply packages/cli/src/__tests__/artificial-panic.test.ts:59]: Yes, once https://github.com/prisma/prisma-private/issues/203 is closed (`prisma@4.10.0`)

File: packages/internals/src/IntrospectionEngine.ts
Comment (Jolg42) [inline packages/internals/src/IntrospectionEngine.ts:274]:
```suggestion
   * @deprecated Use `MigrateEngine.introspect()` instead from `@prisma/migrate` package
```

File: packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts
Comment (janpio) [inline packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:314]: Is this really necessary here? Doesn't `introspect()` take care of that itself?

File: packages/integration-tests/src/__tests__/integration/mariadb/__database.ts
Comment (janpio) [inline packages/integration-tests/src/__tests__/integration/mariadb/__database.ts:28]: Why is this needed?

File: packages/integration-tests/src/__tests__/integration/mysql/__database.ts
Comment (janpio) [inline packages/integration-tests/src/__tests__/integration/mysql/__database.ts:27]: Why is this needed?

File: packages/integration-tests/src/__tests__/integration/postgresql/__database.ts
Comment (janpio) [inline packages/integration-tests/src/__tests__/integration/postgresql/__database.ts:22]: Why is this needed?

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [inline packages/migrate/src/MigrateEngine.ts:48]: Is this only used for Introspection? Then it might make sense to include that in the name maybe. `latestIntrospectedSchema` for example.

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [inline packages/migrate/src/MigrateEngine.ts:153]: Can we already create an issue for this?

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [inline packages/migrate/src/MigrateEngine.ts:507]: Issue?

File: packages/internals/src/__tests__/handlePanic.test.ts
Comment (Jolg42) [inline packages/internals/src/__tests__/handlePanic.test.ts:137]: Not sure what we want to do here, it's still about introspection but the code leaves in the migration engine.

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:172]: Nice improvement that this is a proper error now!

File: packages/internals/src/__tests__/handlePanic.test.ts
Comment (jkomyno) [reply packages/internals/src/__tests__/handlePanic.test.ts:137]: The `RustPanic` in `MigrateEngine` are statically defined as `LIFT_CLI`. Although it's technically introspection-related, this tells us the engine that was used under the hood. Besides, this `ErrorArea` label is of internal-use only

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:174]: This message is not optimal, but I guess unrelated to this PR and also not super important.

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:438]: OT: One day I hope we will figure out what adds these useless spaces all the time and can fix it.

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:174]: Yep, this is unrelated to this PR

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:438]: I agree. Using jest with the `-u` flag causes this, from time to time. Not sure why exactly it happens, though

File: packages/migrate/src/__tests__/introspection/introspection.test.ts
Comment (janpio) [inline packages/migrate/src/__tests__/introspection/introspection.test.ts]: Is this still needed?

File: packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts
Comment (jkomyno) [reply packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts]: I think @Jolg42 ninjed you, I have fixed this already while you were writing this :)

File: packages/integration-tests/src/__tests__/integration/mariadb/__database.ts
Comment (jkomyno) [reply packages/integration-tests/src/__tests__/integration/mariadb/__database.ts:28]: I'll be honest, in retrospect it doesn't seem useful

File: packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts
Comment (janpio) [reply packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts]: I can live with that.

File: packages/cli/src/__tests__/artificial-panic.test.ts
Comment (janpio) [reply packages/cli/src/__tests__/artificial-panic.test.ts:59]: How does that influence it?

**Review by janpio — APPROVED** [review]

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (Jolg42) [inline packages/migrate/src/__tests__/DbPull.test.ts:457]: Unrelated to this PR I guess: I'm not a fan of printing the user and password to the logs 😨 
I'll create an issue for this

Comment (Jolg42) [reply packages/migrate/src/__tests__/DbPull.test.ts:457]: It's only for --url, still undocumented so it's not a blocker

Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:457]: I think we already handle secretation elsewhere. Perhaps this should be a function defined just once in the Rust engines and exported via Wasm. Can you please open an issue for this?

File: packages/integration-tests/src/__tests__/integration/mariadb/__database.ts
Comment (jkomyno) [reply packages/integration-tests/src/__tests__/integration/mariadb/__database.ts:28]: `git blame` tells me that @Jolg42 added it

File: packages/migrate/src/MigrateEngine.ts
Comment (jkomyno) [reply packages/migrate/src/MigrateEngine.ts:48]: Done

File: packages/integration-tests/src/__tests__/integration/mariadb/__database.ts
Comment (Jolg42) [reply packages/integration-tests/src/__tests__/integration/mariadb/__database.ts:28]: Yes it's a nice improvement we already did that for SQL Server only I think, found out when we had the exhaustion of connections on PostrgeSQL

File: packages/migrate/src/__tests__/DbPush.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPush.test.ts:193]: They are behaviorally the same, but the new one was suggested by the Jest extension I'm using in VSCode, which leads me to believe it's the most common "expect" method for that task. It's also compatible with the standard `jasmine` nomenclature, adopted by all other test frameworks. I didn't really notice this change as it was an automated one

File: packages/migrate/src/__tests__/MigrateDiff.test.ts
Comment (jkomyno) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:33]: Same as above

File: packages/migrate/src/MigrateEngine.ts
Comment (jkomyno) [reply packages/migrate/src/MigrateEngine.ts:153]: There's already an issue linked in the comments

**Review by Jolg42 — APPROVED** [review]
Nice! Let's do this 🚀

File: packages/internals/src/__tests__/handlePanic.test.ts
Comment (Jolg42) [reply packages/internals/src/__tests__/handlePanic.test.ts:137]: `INTROSPECTION_CLI` would be better to keep for the meaning of the error was actually triggered from the "introspection part of the CLI"
But since it would be complicated and would not provide any benefit, we can go for `LIFT_CLI`
Technically, we could also remove completely the `ErrorArea` one day (but we need to handle that in the error reporting backend)

File: packages/cli/src/__tests__/artificial-panic.test.ts
Comment (jkomyno) [reply packages/cli/src/__tests__/artificial-panic.test.ts:59]: Besides the actual database version (which is WIP) and the error area (which doesn't matter), this PR will preserve every panic information as before

File: packages/migrate/src/MigrateEngine.ts
Comment (jkomyno) [inline packages/migrate/src/MigrateEngine.ts:507]:
```suggestion
  // TODO: https://github.com/prisma/prisma/issues/17268
```

File: packages/migrate/src/MigrateEngine.ts
Comment (janpio) [reply packages/migrate/src/MigrateEngine.ts:153]: That seemed to be a separate TODO, so I expected a different issue. 
(The other issue is also not really about that TODO, but a requesite for that TODO)

File: packages/cli/src/__tests__/artificial-panic.test.ts
Comment (janpio) [reply packages/cli/src/__tests__/artificial-panic.test.ts:59]: > and the error area (which doesn't matter)

There is literally a filter by error area in the error report tool.

File: packages/migrate/src/__tests__/DbPull.test.ts
Comment (Jolg42) [reply packages/migrate/src/__tests__/DbPull.test.ts:457]: https://github.com/prisma/prisma/issues/17286

**Comment by janpio** [comment]
Let's create a public `prisma/prisma` issue for this that will show up in our release notes so people can learn about this.

**Comment by Jolg42** [comment]
Note: first appeared in `4.9.0-dev.55`

**Pull request merged**
